### PR TITLE
Experience replay

### PR DESCRIPTION
**Pull request by mesjou**

### CODE REVIEW

**Comment by mesjou** [comment]
@SimonHashtag implementation of experience replay. I also did a important change in the environment. Changed reward to be always between +1 and -1. If we want the negative reward to stay at -10 in the environment we can also scale everything. This  quick fix improved performance significantly

File: lifecycle/agents/dqn.py [hunk 1]
```diff
@@ -65,49 +90,83 @@ def train(self, obs: Dict, next_obs: Dict, act: float, rew: float, done: bool):
...
+            if self._double_dqn:
+                local_q = self._local_network(tf.convert_to_tensor(next_states, dtype=tf.float32),
+                                               tf.convert_to_tensor(next_action_mask, dtype=tf.float32))
+                next_action = tf.argmax(local_q, axis=1)
+                target_value = tf.reduce_sum(tf.one_hot(next_action, self._n_actions) * target_q, axis=1)
+            else:
+                next_action = tf.argmax(target_q, axis=1)
+                target_value = tf.reduce_sum(tf.one_hot(next_action, self._n_actions) * target_q, axis=1)
+
+            if self._false_dones:
```
Comment (SimonHashtag) [inline lifecycle/agents/dqn.py:133]: Does configuration with _false_dones = True and n_step = 1 work as good as or better than a correct specification?

File: lifecycle/agents/dqn.py [hunk 2]
```diff
@@ -65,49 +90,83 @@ def train(self, obs: Dict, next_obs: Dict, act: float, rew: float, done: bool):
...
+                target_value = self._gamma * target_value + rewards
+            else:
+                target_value = (1 - dones) * self._gamma * target_value + rewards
+
+            main_q = self._local_network(tf.convert_to_tensor(states, dtype=tf.float32),
+                                    tf.convert_to_tensor(action_mask, dtype=tf.float32))
+            main_value = tf.reduce_sum(tf.one_hot(actions, self._n_actions) * main_q, axis=1)
+
+            td_error = main_value - target_value
+            error = tf.square(td_error) * 0.5
```
Comment (SimonHashtag) [inline lifecycle/agents/dqn.py:143]: Why is this scaled with 0.5?

File: lifecycle/agents/dqn.py [hunk 3]
```diff
@@ -65,49 +90,83 @@ def train(self, obs: Dict, next_obs: Dict, act: float, rew: float, done: bool):
...
+        if self._clip_grads:
+            gradients = [(tf.clip_by_value(grad, -1.0, 1.0)) for grad in gradients]
+
+        if self._prioritized_replay:
+            new_priorities = np.abs(td_error) + 1e-6  # todo change to editable?
+            self._replay_buffer.update_priorities(batch["index"], new_priorities)
+
+        self._optimizer.apply_gradients(zip(gradients, dqn_variable))
+
+        return error
```
Comment (SimonHashtag) [inline lifecycle/agents/dqn.py:157]: More comments for section l. 113-157 might be helpful

File: lifecycle/train_dqn.py [hunk 4]
```diff
@@ -102,14 +133,18 @@ def run_dqn(nb_episode: int = 500):
 
 def test(Agent: DQN, env: gym.Env):
     """Return Q-Function and optimal policy of trained neural network."""
-    states = range(env.get_number_states())
-    action_masks = [env.get_action_mask(state) for state in states]
-    q = np.vstack(tuple(np.array(Agent.predict(np.array(s), mask)) for s, mask in zip(states, action_masks)))
+    states = np.arange(0, env.get_number_states())
+    action_masks = [np.ones_like(env.get_action_mask(state)) for state in states]
+    # action_masks = [env.get_action_mask(state) for state in states]
+    states = np.array([[state, state ** 1/2, state ** 1/3, 0.1, 0.2] for state in states])
```
Comment (SimonHashtag) [inline lifecycle/train_dqn.py:139]: Why is this array necessary?

**Review by SimonHashtag — COMMENTED** [review]
Good job :)
Added some questions as comments.

File: lifecycle/train_dqn.py [hunk 5]
```diff
@@ -34,30 +41,49 @@ def run_dqn(nb_episode: int = 500):
         lr_schedule = initial_learning_rate
 
     # init policies
-    env = SimpleConsEnv()
+    env = SimpleConsEnvDouble()
+    # todo add more sophisticated observatiopn wrapper that always transform observationn normal with mean 0
+    env = TransformObservation(env, lambda obs: obs / 50.0 + 0.1)
```
Comment (SimonHashtag) [inline lifecycle/train_dqn.py:46]: Scale states so that the range from -1 to 1.
lambda obs: obs / 50.0 - 1.0

File: lifecycle/agents/dqn.py [hunk 6]
```diff
@@ -32,21 +39,39 @@ def __init__(
         self._local_network = nn(observation_space, hidden_units, n_actions)
         self.target_network = nn(observation_space, hidden_units, n_actions)
         self._optimizer = tf.optimizers.Adam(lr)  # could be fixed lr or lr schedule
+        # todo remove later
+        self._local_network.run_eagerly = True  # for debugging
```
[the remark this answers is not in the document]
Comment (mesjou) [reply lifecycle/agents/dqn.py:43]: no its nice for debugging, if you call it it is possible to stop the program even inside tf functions. we can remove it later or set to false if we need good performance and code works

File: lifecycle/agents/dqn.py (hunk 1 above)
Comment (mesjou) [reply lifecycle/agents/dqn.py:133]: yes I think it works better at least if we want to come close the dynamic programming solution. If periods end this changes the problem from continuous to finite and this somehow makes it worse for the algorithm to learn. I think it could be that at random points for him (because t is not in the state) the transition to the next period only includes the reward but not the q-values of the next states. So he must also learn with which probability this will happen in which state. Without this done he always takes next period q-value into account

**Review by SimonHashtag — APPROVED** [review]

**Comment by SimonHashtag** [comment]
Changes:
- Experience priorization replay
- Reward and states (during training) scaled to [-1, 1] -> improves performance significantly
- env wrapping
- gradient clipper
